src/components/vega-lite/index.tsx

```diff
@@ -50,31 +55,69 @@ export class VegaLite extends React.PureComponent<VegaLiteProps, any> {
     //   }
     // };
     const {logger} = this.props;
+    const vlSpec = this.props.spec;
     try {
       const spec = vl.compile(vlSpec, logger).spec;
       const runtime = vega.parse(spec, vlSpec.config);
-      const view = new vega.View(runtime)
+      this.view = new vega.View(runtime)
         .logLevel(vega.Warn)
         .initialize(this.refs[CHART_REF] as any)
         .renderer(this.props.renderer || 'svg')
-        .hover()
-        .run();
-      vegaTooltip.vega(view);
+        .hover();
+      vegaTooltip.vega(this.view);
+      this.bindData();
     } catch (err) {
       logger.error(err);
     }
   }
 
   protected componentDidMount() {
-    this.renderVega(this.props.spec);
+    this.updateSpec();
+    this.runView();
+  }
+
+  protected componentWillUpdate() {
+    this.size = this.getChartSize();
   }
 
-  protected componentWillReceiveProps(nextProps: VegaLiteProps) {
-    if (this.props.spec !== nextProps.spec) {
-      setTimeout(() => {
-        this.renderVega(nextProps.spec);
-      });
+  protected componentDidUpdate(prevProps: VegaLiteProps, prevState: {}) {
+    if (prevProps.spec !== this.props.spec) {
+      const chart = this.refs[CHART_REF] as HTMLElement;
+      chart.style.width = this.size[0] + 'px';
+      chart.style.height = (this.size[1] + 3) + 'px';
+      // add extra 3px to height so chart can contain the svg/canvas plot
```

Comment:
Because the size the svg is slightly smaller than its container div chart. I inspected the element, and the height of svg is 3px smaller than its div.